src/GMCPAuthenticator.cpp

```diff
@@ -0,0 +1,108 @@
+/***************************************************************************
+ *   Copyright (C) 2024 by Vadim Peretokin - vperetokin@gmail.com          *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
+ ***************************************************************************/
+
+#include "GMCPAuthenticator.h"
+
+#include "Host.h"
+#include "ctelnet.h"
+#include <QDebug>
+
+GMCPAuthenticator::GMCPAuthenticator(Host* pHost) : mpHost(pHost) {}
+
+void GMCPAuthenticator::saveSupportsSet(const QString& data)
+{
+    auto jsonDoc = QJsonDocument::fromJson(data.toUtf8());
+    auto jsonObj = jsonDoc.object();
+
+    if (jsonObj.contains("type")) {
+        QJsonArray typesArray = jsonObj["type"].toArray();
+        for (const auto& type : typesArray) {
+            mSupportedAuthTypes.append(type.toString());
+        }
+    }
+
+    qDebug() << "Supported auth types:" << mSupportedAuthTypes;
+}
+
+void GMCPAuthenticator::sendCredentials()
+{
+    auto character = mpHost->getLogin();
+    auto password = mpHost->getPass();
+    if (character.isEmpty() || password.isEmpty()) {
+        qDebug() << "No login or password set in connection settings. Cannot authenticate with GMCP.";
+        return;
+    }
+
+    QJsonObject credentials;
+    credentials["account"] = character;
+    credentials["password"] = password;
+
+    QJsonDocument doc(credentials);
+    QString gmcpMessage = doc.toJson(QJsonDocument::Compact);
```

Comment:
Could this be updated to support the optional part of the spec about sending an empty hash in the reply if the client doesn't have both username and password?  I guess something like this should work:

```suggestion
void GMCPAuthenticator::sendCredentials()
{
    auto character = mpHost->getLogin();
    auto password = mpHost->getPass();
    QJsonObject credentials;
    if (!character.isEmpty() && !password.isEmpty()) {
        credentials["account"] = character;
        credentials["password"] = password;
    }
    QJsonDocument doc(credentials);
    QString gmcpMessage = doc.toJson(QJsonDocument::Compact);
```